File: November 2025 GFG SOLUTION/November-22.py
class Solution:
    def minConnect(self, V, edges):
        comp, e = V, len(edges)
        if e < V - 1: return -1
        p = list(range(V))
        def find(x): 
            if p[x] != x: p[x] = find(p[x])
            return p[x]
        for u, v in edges:
            x, y = find(u), find(v)
            if x != y: p[x] = y; comp -= 1
        return comp - 1

File: November 2025 GFG SOLUTION/November-22.java
class Solution {
    public int minConnect(int V, int[][] edges) {
        int comp = V, e = edges.length;
        if (e < V - 1) return -1;
        int[] p = new int[V];
        for (int i = 0; i < V; i++) p[i] = i;
        for (int[] ed : edges) {
            int x = find(p, ed[0]), y = find(p, ed[1]);
            if (x != y) { p[x] = y; --comp; }
        }
        return comp - 1;
    }
    int find(int[] p, int x) { 
       return p[x] == x ? x : (p[x] = find(p, p[x])); 
    }
}
